flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoOperatorTracker.java

```diff
@@ -0,0 +1,403 @@
+
+        /**
+         * Constructs a new {@link ThreadInfoOperatorTracker}.
+         *
+         * @return a new {@link ThreadInfoOperatorTracker} instance.
+         */
+        public ThreadInfoOperatorTracker<T> build() {
+            return new ThreadInfoOperatorTracker<>(
+                    coordinator,
+                    resourceManagerGatewayRetriever,
+                    createStatsFn,
+                    executor,
+                    cleanUpInterval,
+                    numSamples,
+                    statsRefreshInterval,
+                    delayBetweenSamples,
+                    maxThreadInfoDepth);
+        }
+    }
+
+    private static final Logger LOG = LoggerFactory.getLogger(ThreadInfoOperatorTracker.class);
+
+    /** Lock guarding trigger operations. */
+    private final Object lock = new Object();
+
+    private final ThreadInfoRequestCoordinator coordinator;
+
+    private final Function<OperatorThreadInfoStats, T> createStatsFn;
+
+    private final ExecutorService executor;
+
+    private final GatewayRetriever<ResourceManagerGateway> resourceManagerGatewayRetriever;
+
+    /**
+     * Completed stats. Important: Job vertex IDs need to be scoped by job ID, because they are
+     * potentially constant across runs messing up the cached data.
+     */
+    private final Cache<AccessExecutionJobVertex, T> operatorStatsCache;
+
+    /**
+     * Pending in progress stats. Important: Job vertex IDs need to be scoped by job ID, because
+     * they are potentially constant across runs messing up the cached data.
+     */
+    private final Set<AccessExecutionJobVertex> pendingStats = new HashSet<>();
+
+    private final int numSamples;
+
+    private final int statsRefreshInterval;
+
+    private final Time delayBetweenSamples;
+
+    private final int maxThreadInfoDepth;
+
+    // Used for testing purposes
+    private final CompletableFuture<Void> resultAvailableFuture = new CompletableFuture<>();
+
+    /** Flag indicating whether the stats tracker has been shut down. */
+    private boolean shutDown;
+
+    private ThreadInfoOperatorTracker(
+            ThreadInfoRequestCoordinator coordinator,
+            GatewayRetriever<ResourceManagerGateway> resourceManagerGatewayRetriever,
+            Function<OperatorThreadInfoStats, T> createStatsFn,
+            ExecutorService executor,
+            int cleanUpInterval,
+            int numSamples,
+            int statsRefreshInterval,
+            Time delayBetweenSamples,
+            int maxStackTraceDepth) {
+
+        this.coordinator = checkNotNull(coordinator, "Thread info samples coordinator");
+        this.resourceManagerGatewayRetriever =
+                checkNotNull(resourceManagerGatewayRetriever, "Gateway retriever");
+        this.createStatsFn = checkNotNull(createStatsFn, "Create stats function");
+        this.executor = checkNotNull(executor, "Scheduled executor");
+
+        checkArgument(cleanUpInterval >= 0, "Clean up interval");
```

Comment:
Very good question. Guava will effectively disable caching if the clean-up interval is 0 [1]:
```
When duration is zero, this method hands off to maximumSize(0), ignoring any otherwise-specificed maximum size or weight. This can be useful in testing, or to disable caching temporarily without a code change.
``` https://guava.dev/releases/18.0/api/docs/com/google/common/cache/CacheBuilder.html#expireAfterAccess(long,%20java.util.concurrent.TimeUnit)

This could be useful, if we want to give users more control from the UI with regards to how often to force resampling of the results. As this functionality is currently not supported by the UI, it probably makes sense to only allow values >0 for now. What do you think?

Reply:
And probably also the same for the other parameters that currently accept 0s.